# -----------------------------------------------------------------------------
#
# Plot distribution of relative lengths of species-yeast orthologs and %identities
#
# Jason Jiang - Created: 2022/08/29
#               Last edited: 2022/08/31
#
# Reinke Lab - Microsporidia Orthologs Project
#
# Goal: Plot the distribution of species-yeast ortholog pairs with their relative
# lengths and percent identities when aligned.
#
# Thanks to Brandon Murareanu for this RScript layout
#
#
# ------------------------------------------------------------------------------

suppressMessages(library(tidyverse))
suppressMessages(library(svglite))

################################################################################

main <- function() {
  # ---------------------------------------------------------------------------
  # Command line args:
  #   $1 = dataframe of species-yeast orthologs and theirlengths
  #
  #   $2 = hashtable mapping species genes to their alignments w/ their yeast
  #   orthologs and percent identities (from make_ortholog_alignments_hashtable.R)
  #
  #   $3 = list of essential yeast genes
  #
  #   $4 = directory to save resulting plots in
  #
  # ---------------------------------------------------------------------------
  args <- commandArgs(trailingOnly = T)
  orthogroups <- read_csv(args[1], show_col_types = F)
  ortholog_alignments <- read_rds(args[2])
  essential_yeast_genes <- read_lines(args[3])
  out <- args[4]
  
  # TODO - replace hard-coded arguments w/ variable args
  orthogroups <- read_csv('../../data/orthogroups.csv', show_col_types = F)
  ortholog_alignments <- read_rds('../../resources/ortholog_alignments.rds')
  essential_yeast_genes <- read_lines('../../data/essential_yeast_genes.txt')
  out <- '../../results/lengths_vs_percent_identities'
  
  # annotate each species-yeast ortholog pair w/ whether the yeast ortholog is
  # essential
  #
  # calculate percent id for each species-yeast single-copy ortholog pair using
  # their alignments
  orthogroups <- orthogroups %>%
    mutate(essential = yeast_ortholog %in% essential_yeast_genes) %>%
    calculate_ortholog_pids(ortholog_alignments)
  
  # split orthogroups dataframe into groups of species
  microsp_orthogroups <- filter(orthogroups, is_microsp, species != 'R_allo')
  rozella_orthogroups <- filter(orthogroups, species == 'R_allo')
  outgroup_orthogroups <- filter(orthogroups, !is_microsp)
  
  # plot ortholog lengths vs percent identities for each group of species
  dir.create(out, showWarnings = F)
  plot_lengths_vs_percent_id(microsp_orthogroups, 'Microsporidia', out)
  plot_lengths_vs_percent_id(rozella_orthogroups, 'Rozella allomycis', out)
  plot_lengths_vs_percent_id(outgroup_orthogroups, 'Outgroup species', out)
}

################################################################################

## Helper functions

calculate_ortholog_pids <- function(orthogroups, ortholog_alignments) {
  # ---------------------------------------------------------------------------
  # Get percent id for each species-yeast ortholog pair, adding it as a column
  # to the orthogroups dataframe
  #
  # Arguments:
  #   orthogroups: dataframe of single-copy species-yeast orthologs, generated
  #   by 'XXX.R'
  #
  #   ortholog_alignments: hashtable of species ortholog names and their alignments
  #   to their yeast orthologs, generated by 'YYY.R'
  #
  # ---------------------------------------------------------------------------
  return(
    orthogroups %>%
      select(orthogroup, essential, species, is_microsp, yeast_ortholog,
             species_ortholog, yeast_len, species_len) %>%
      mutate(length_ratio = species_len / yeast_len) %>%
      rowwise() %>%
      mutate(percent_id = ortholog_alignments[[species_ortholog]]$percent_id)
  )
}


plot_lengths_vs_percent_id <- function(orthogroups, name, out) {
  # ---------------------------------------------------------------------------
  # Saves a plot of relative ortholog lengths of the species to its yeast
  # orthologs, versus the percent identities between the species-yeast
  # ortholog pairs.
  #
  # Arguments:
  #   orthogroups: dataframe containing ortholog pair lengths and percent ids
  #
  #   name: name of the group of species contained in orthogroups dataframe
  #
  #   out: directory to save the resulting plot in
  #
  # ---------------------------------------------------------------------------
  plot <- ggplot(orthogroups, aes(y = percent_id, x = length_ratio)) +
    geom_point(aes(colour = factor(essential)), alpha = 0.5) +
    geom_rug(alpha = 0.5, aes(colour = factor(essential))) +
    labs(x = 'Length of species ortholog / Length of yeast ortholog',
         y = '% Identity (Identical positions / Aligned positions)',
         title = str_c(name, '\nn = ', nrow(orthogroups), ' ortholog pairs\n',
                       length(unique(orthogroups$orthogroup)), ' orthogroups'),
         color = 'Essential in yeast?') +
    theme_bw() +
    theme(axis.title = element_text(color = 'black', size = 18),
          axis.text = element_text(size = 18),
          title = element_text(color = 'black', size = 18),
          legend.text = element_text(size = 14))
  
  ggsave(plot = plot, filename = str_c(out, '/', name, '.svg'),
         dpi = 600, units = 'in', width = 9, height = 7.33)
}

################################################################################

main()